collector/lolperf/benchlib/src/measure.rs

```diff
@@ -0,0 +1,99 @@
+use crate::messages::BenchmarkResult;
+use perf_event::events::Hardware;
+use perf_event::{Builder, Counter, Group};
+use std::time::Instant;
+
+struct Counters {
+    cycles: Counter,
+    instructions: Counter,
+    branch_misses: Counter,
+    cache_misses: Counter,
+    cache_references: Counter,
+}
+
+/// (Micro-)benchmarks a single function.
+pub fn benchmark<F: FnOnce() -> R, R>(
+    name: &'static str,
+    func: F,
+) -> anyhow::Result<BenchmarkResult> {
+    let mut group = create_group()?;
+    let counters = prepare_counters(&mut group)?;
+
+    let start = Instant::now();
+
+    // Do not act on the return value to avoid including the branch in the measurement
+    let enable_ret = group.enable();
+    let output = func();
+    group.disable()?;
+
+    let duration = start.elapsed();
+
+    // Check if we have succeeded before
+    enable_ret?;
+
+    // Try to avoid optimizing the result out
+    black_box(output);
+
+    let measurement = group.read()?;
+
+    let result = BenchmarkResult {
+        name,
+        cycles: measurement[&counters.cycles],
+        instructions: measurement[&counters.instructions],
+        branch_misses: measurement[&counters.branch_misses],
+        cache_misses: measurement[&counters.cache_misses],
+        cache_references: measurement[&counters.cache_references],
+        wall_time: duration,
+    };
+    Ok(result)
+}
+
+pub fn create_group() -> anyhow::Result<Group> {
+    match Group::new() {
+        Ok(group) => Ok(group),
+        Err(error) => {
+            let path = "/proc/sys/kernel/perf_event_paranoid";
+            let level = std::fs::read_to_string(path).unwrap_or_else(|_| "unknown".to_string());
+            let level = level.trim();
+            Err(anyhow::anyhow!(
+                "Cannot create perf_event group ({:?}). Current value of {} is {}. Try lowering it.",
```

Comment:
More info about what command to run to change it?

Reply:
Good point. It's not really a specific command, you just have to write the value into the given file path (with .e.g. `echo`). I'll add the hint.